### PR TITLE
Document why Vec::with_capacity is useful

### PR DESCRIPTION
**Pull request by jyn514**
The existing docs showed how to use it, but not why.

I wrote this up for work and figured I may as well make a PR since I'd done all the hard work.

### CODE REVIEW

**Comment by jyn514** [comment]
Let me know if I should remove the references to `hyperfine` - the benchmark seemed useful but I understand if the standard library should be impartial to rust community tools.

File: library/alloc/src/vec.rs [hunk 1]
```diff
@@ -333,6 +333,55 @@ impl<T> Vec<T> {
...
+    /// copying the entire buffer - the last one will require copying 67 MB.
+    ///
+    /// On my machine this runs in 148 ms:
+    ///
+    /// ```text
+    /// $ hyperfine ./allocate --warmup 1
+    /// Benchmark #1: ./allocate
+    ///   Time (mean ± σ):     147.6 ms ±   0.5 ms    [User: 121.9 ms, System: 25.8 ms]
+    ///   Range (min … max):   146.7 ms … 148.7 ms    20 runs
+    /// ```
```
Comment (LukasKalbertodt) [inline library/alloc/src/vec.rs:363]: I think I would completely remove the specific benchmarks.

- "on my machine": I don't think std has a formulation like that anywhere. 
- I don't think `hyperfine` is the best tool for these kinds of benchmarks. Quickly looking at its README, it seems to be mostly for benchmark command line programs. It apparently starts the payload as a normal command, which has quite some overhead. Might not matter for this particular benchmark, but using the built-in benchmark utility would make more sense.

I think the information "it's faster" and the explanation about doubling the capacity is sufficient.

**Review by LukasKalbertodt — COMMENTED** [review]
Thanks for your PR! 

I'm a bit unsure about this change. I added a few inline comments that would need to be resolved, but in general I am not sure if this is the best place to explain this.

Maybe we should rather add your code example + explanation with specific numbers to the section `#capacity-and-reallocation` in the type-level docs. And then add only one or two sentences to `with_capacity` and `reserve`.

**Comment by LukasKalbertodt** [comment]
Maybe someone from @rust-lang/docs could take this PR or help out?

**Comment by steveklabnik** [comment]
I pretty much agree with all of your points, @LukasKalbertodt . I think adding a section that explains this is a great idea, but this needs some tweaks.

**Comment by jyn514** [comment]
Hmm, your points make sense to me - I completely missed that there was already a 'Capacity and reallocation' section. Do you think this is still worthwhile without the benchmarks and given the info already present? Other than the specific numbers and the code examples I don't know that this contributes much.

**Comment by LukasKalbertodt** [comment]
I think it's still useful to add a bit of information (what I mentioned above). But I can also understand if you want to close the PR as what we want to merge is quite different from your original text. You decide :)

**Comment by jyn514** [comment]
I don't expect to get around to this any time soon; if someone else wants to pick it up go for it :)

**Pull request closed without merging**